Bug 1082693 - Fetch bug summaries from bugzilla for use in tooltips

KWierso opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/764)
<!-- Reviewable:end -->

edmorley commented:
I like this overall - though I wonder since we're already using jquery whether it makes sense to use that instead of the fetch API, for greater compatibility? There also seems to be a lot of boilerplate with the fetch API that you save with jquery.

Also, I think we need to more aggressively cache the bug lookups + also add a slight delay, so that moving the mouse around the screen doesn't trigger 50 requests to Bugzilla's API (a la: https://bugzilla.mozilla.org/show_bug.cgi?id=799614).

KWierso commented:
How would you want to cache them? At the moment, it only fetches a bug's summary from bugzilla once per page load, so moving the mouse over a bug link, away from it, and back to it will only make one request to bugzilla (unless you do it really quickly before the first network request can finish). If more than one commit references the same bug number, it WILL make a network request for each one.

Should I just make an array on the user's browser or something to store each bug's summary and try searching in it before making a network request? That would still be blown away if the user refreshes the page, but I guess that's good since the summary/status could have changed.

Or what about storing that array more persistently in localStorage or indexedDB or something along with a timestamp so we don't hit the network unless it's either not stored at all or not checked since a certain time?

edmorley commented:
Once per pageload was all I was after; if it does that at present, then great :-) (hadn't tested locally yet)

edmorley commented:
Oh re-read. Might be nice to fix the "multiple bug numbers" case - particularly since we may re-use this code elsewhere in the UI, where bugs may repeat more often.

KWierso commented:
Can do. 

Haven't tested using jquery for this. I tried using $http but I haven't figured out how to inject it into the function.

edmorley commented:
This is getting close :-)
The caching works well - the only thing is that some of the time the tooltip doesn't update and stays as "bugzilla.mozilla.org" - possibly when the Bugzilla API response takes a bit longer.

Also, could you rebase on latest master? To test this with runserver I was getting errors until I downgraded django-rest-framework, since my vagrant environment has the deps from master installed.

KWierso commented:
Rebased to master and fixed an eslint issue.

I think the tooltip still showing "bugherder.mozilla.org" if the BMO request takes too long is an inherent "problem" with tooltips (or at least, that's what I get from searching for this and finding http://stackoverflow.com/questions/23549625/keep-tooltip-title-attribute-updated-while-mouseover )

If you move the cursor off of the bug number and back on, it will show the correct bug information.

KWierso commented:
Bug 593020 seems relevant.

KWierso commented:
So this adds the delay, but now it's showing the commit message instead of the bugzilla information in the tooltip all the time due to Bug 593020.

I think the only way we'll get around it is to use something like jquery-ui or angular tooltips to override the native behavior.

KWierso commented:
Also need to fix what happens with security bugs, since those return a 401 error and will continuously try to fetch the summary every time the mouseover hover happens...

camd commented on ui/js/filters.js:
```diff
@@ -144,3 +147,61 @@ treeherder.filter('alertStatus', [
+
+    if(bugID) {
+        // Fetch the bug information from bugzilla and store it locally
+        $.getJSON(bugURL)
```
Instead of using a function like this, we should create a ``Bugzilla`` model to call these endpoints.  It may well come in handy in other parts of our code-base, too.

wlach replied:
Yeah there's a lot of places in Perfherder that could benefit from showing a bug summary.

camd commented:
I mentioned in a bug comment, but I'll mention here, too.  You may check out http://angular-ui.github.io/bootstrap/#/popover as an alternative to using conventional tooltips.  You could likely set the value shown to an angular binding like ``{{ bugInfo.blah.blah }}`` and you wouldn't have to un-hover and re-hover to get the value.  it'd just show when the ajax returns.

KWierso commented:
Starting to work on this. I've added the ability to fetch a bug's information by requesting it from the TH server. 

At the moment, the only thing the endpoint does is take any "include_fields" param from the request and passes it along in the request to Bugzilla.

I had to create that blank treeherder/templates/debug_toolbar_base.html file because requests to the new get_bug endpoint were failing until I added it. It seems to work just fine with it empty, and I'm open to suggestions on what I'm doing wrong that makes it thing I need that file.

The new endpoint does no error handling whatsoever at the moment, I'll fix that up before going forward.

I'm a bit confused on why I have to put the bugID before /get_bug/. Am I doing something wrong in setting this up? I would've thought it'd be in the form /get_bug/bugID, but that fails.


Regarding adding a Bugzilla model to do this, how can I do that from filters.js? These bug links are only created within the filter, and I don't seem to have $scope or anything in there...

edmorley commented:
Ah - when Cameron said "use a Bugzilla model" he doesn't mean fetch from the server, he means create a new file in the UI directory to group together common Bugzilla functions, like we do in treestatus.js.

wlach commented on treeherder/webapp/api/bugzilla.py:
```diff
@@ -48,3 +48,25 @@ def create_bug(self, request):
+
+        response = make_request(url, method='GET', headers=headers)
+        # Needs error handling, but I don't care right now
+        return Response(response.json())
```
I was looking at this in the context of something else and I have to ask, why are we trying to add an endpoint to the treeherder server? Can't we just query bugzilla directly and avoid a layer of indirection and load on treeherder's part?

wlach replied:
@edmorley pointed out that it was originally written this way, and only changed to a misunderstanding. Please disregard the above :)

wlach commented on ui/js/filters.js:
```diff
@@ -144,3 +147,61 @@ treeherder.filter('alertStatus', [
+
+function getBugInfo(bugLink) {
+    var bugID = bugLink.getAttribute("data-bugid");
+    var bugURL = '/api/bugzilla/' + bugID + '/get_bug/?include_fields=status,summary';
```
So if we're going to query bugzilla directly, I think you can just replace this line with:

```js
var bugURL = 'https://bugzilla.mozilla.org/rest/bug/' + bugID + '/get_bug/?include_fields=status,summary';
```

and it should just work. This has the additional bonus of letting you test any changes to the feature completely on the client side.

KWierso commented:
I don't see myself get to this anytime soon, so I'm gonna close it out in the interest of making the number of open pull requests go down.
